test_utils/git.py

```diff
@@ -0,0 +1,117 @@
+#
+# Copyright(c) 2019-2022 Intel Corporation
+# Copyright(c) 2024 Huawei Technologies Co., Ltd.
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+import os
+
+from core.test_run import TestRun
+from connection.local_executor import LocalExecutor
+from test_utils.output import CmdException
+
+
+def get_submodules_paths(from_dut: bool = False):
+    executor = TestRun.executor if from_dut else LocalExecutor()
+    repo_path = TestRun.usr.working_dir if from_dut else TestRun.usr.repo_dir
+    git_params = "config --file .gitmodules --get-regexp path | cut -d' ' -f2"
+
+    output = executor.run(f"git -C {repo_path} {git_params}")
+    if output.exit_code != 0:
+        raise CmdException("Failed to get submodules paths", output)
+
+    return output.stdout.splitlines()
+
+
+def get_repo_files(
+    branch: str = "HEAD",
+    with_submodules: bool = True,
+    with_dirs: bool = False,
+    from_dut: bool = False,
+):
+    executor = TestRun.executor if from_dut else LocalExecutor()
+    repo_path = TestRun.usr.working_dir if from_dut else TestRun.usr.repo_dir
+    git_params = f"ls-tree -r --name-only --full-tree {branch}"
+
+    output = executor.run(f"git -C {repo_path} {git_params}")
+    if output.exit_code != 0:
+        raise CmdException("Failed to get repo files list", output)
+
+    files = output.stdout.splitlines()
+
+    if with_submodules:
+        for subm_path in get_submodules_paths(from_dut):
+            output = executor.run(f"git -C {os.path.join(repo_path, subm_path)} {git_params}")
+            if output.exit_code != 0:
+                raise CmdException(f"Failed to get {subm_path} submodule repo files list", output)
+
+            subm_files = [os.path.join(subm_path, file) for file in output.stdout.splitlines()]
+            files.extend(subm_files)
+
+    if with_dirs:
+        # use set() to get unique values only
+        dirs = set(os.path.dirname(file) for file in files)
+        files.extend(dirs)
+
+    # change to absolute paths and remove empty values
+    files = [os.path.realpath(os.path.join(repo_path, file)) for file in files if file]
+
+    return files
```

Comment:
How about extracting the listing files functionality into a separate function?
```

def __get_repo_files(branch, repo_path, executor, with_dirs):
    git_params = f"ls-tree -r --name-only --full-tree {branch}"

    output = executor.run(f"git -C {os.path.join(repo_path)} {git_params}")
    if output.exit_code != 0:
        raise CmdException(f"Failed to get {repo_path} repo files list", output)

    files = [os.path.realpath(os.path.join(repo_path, file)) for file in output.stdout.splitlines()]

    if with_dirs:
        dirs = set(os.path.dirname(file) for file in files)
        files.extend(dirs)

    return files

def get_repo_files(
    branch: str = "HEAD",
    with_submodules: bool = True,
    with_dirs: bool = False,
    from_dut: bool = False,
):
    executor, repo_path = __get_executor_and_repo_path()

    repos_to_list = [repo_path]

    if submodules:
        repos_to_list += get_submodules_paths(from_dut)

    files_lists = [file_list for file_list in __get_repo_files(branch, repo_path, executor)]
    # At this point we have a list of lists which must be flattened
    files = list(iterools.chain.from_iterable(files_lists)) 

    return files
```

Reply:
Why not use https://gitpython.readthedocs.io/en/stable/index.html ?

Reply:
> Why not use https://gitpython.readthedocs.io/en/stable/index.html ?

Because it will work only for local repo, and not if we want to work on remote repo from dut